Resolves a possible memory leak in ensure_pdata

lexicalunit opened the pull request:
## What does this PR do?

Resolves https://github.com/nanodbc/nanodbc/issues/413

I also re-ran formatting tools on `README.md` and updated it a bit, it was out of date in the Docker section-- which I was using when trying to get started working on this PR. Incidentally, the `Dockerfile` also was very out of date and needed to be updated so that I could actually build things. I've included these changes in this PR as well because I'm lazy 😅

mloskot commented on nanodbc/nanodbc.cpp:
```diff
@@ -4717,7 +4717,10 @@
 
     if (!success(rc))
+    {
+        delete buffer;
```
Wouldn't it be better to replace `new` with `unique_ptr` all together and make the code scanner happier?

lexicalunit replied:
Yeah! Probably for the best.

mloskot reviewed (changes requested):
@lexicalunit Thanks for chiming in

lexicalunit commented:
@mloskot I've updated to use a `std::unique_ptr` at the site of the initial memory allocation to avoid the possibility of a memory leak. I also noticed that the `style.sh` script was failing for me in the Docker container so I updated it too.

lexicalunit commented on Dockerfile:
```diff
@@ -1,48 +1,70 @@
-FROM ubuntu:22.10
+FROM ubuntu:22.04
```
22.10 isn't supported for ODBC 18 install: https://learn.microsoft.com/en-us/sql/connect/odbc/linux-mac/installing-the-microsoft-odbc-driver-for-sql-server?view=sql-server-ver16&tabs=ubuntu18-install%2Calpine17-install%2Cdebian8-install%2Credhat7-13-install%2Crhel7-offline

lexicalunit commented on Dockerfile:
```diff
@@ -1,48 +1,70 @@
+    && apt-get -qy install --no-upgrade --no-install-recommends \
+    apt-transport-https \
+    apt-utils \
+    clang-format-15 \
```
Added so that we can run formatting tools inside the contianer.

lexicalunit commented on Dockerfile:
```diff
@@ -1,48 +1,70 @@
+    curl \
+    g++ \
+    git \
+    gpg-agent \
```
Added so that we can install MS keys to apt lists.

lexicalunit commented on Dockerfile:
```diff
@@ -1,48 +1,70 @@
+    && tar -xzf mysql-connector-odbc.tar.gz \
+    && rm mysql-connector-odbc.tar.gz \
+    && mv mysql-connector-odbc* /opt/mysql-connector-odbc \
+    && /opt/mysql-connector-odbc/bin/myodbc-installer -d -a -n "MySQL ODBC 8.4 ANSI Driver" -t "DRIVER=/opt/mysql-connector-odbc/lib/libmyodbc8a.so;" \
+    && /opt/mysql-connector-odbc/bin/myodbc-installer -d -a -n "MySQL ODBC 8.4 Unicode Driver" -t "DRIVER=/opt/mysql-connector-odbc/lib/libmyodbc8w.so;"
```
The `DRIVER` paths were incorrect before this PR. I'm not sure this ever worked.

lexicalunit commented on Dockerfile:
```diff
@@ -1,48 +1,70 @@
- && /opt/mysql-connector-odbc-5.3.13-linux-ubuntu17.04-x86-64bit/bin/myodbc-installer -d -a -n "MySQL ODBC 5.3 ANSI Driver" -t "DRIVER=/usr/lib/x86_64-linux-gnu/odbc/libmyodbc5a.so;" \
- && /opt/mysql-connector-odbc-5.3.13-linux-ubuntu17.04-x86-64bit/bin/myodbc-installer -d -a -n "MySQL ODBC 5.3 Unicode Driver" -t "DRIVER=/usr/lib/x86_64-linux-gnu/odbc/libmyodbc5w.so;" \
- && git clone https://github.com/nanodbc/nanodbc.git /opt/nanodbc && mkdir -p /opt/nanodbc/build
+# NOTE: To install latest version of CMake on Ubuntu, see: https://apt.kitware.com/
+RUN curl -SL https://github.com/Kitware/CMake/releases/download/v3.30.1/cmake-3.30.1.tar.gz -o cmake.tar.gz \
+    && tar -xzf cmake.tar.gz \
+    && rm cmake.tar.gz \
+    && mv cmake-* /opt/cmake
+WORKDIR /opt/cmake
+RUN ./bootstrap \
+    && make -j"$(nproc)" \
+    && make install
```
Since the CMake files were updated to use features of a newer version of CMake, I needed to manually update CMake to the latest to get the build to work correctly in the container.

lexicalunit commented on README.md:
```diff
@@ -91,13 +88,13 @@ cd path/to/nanodbc/repository
+make           # creates shared library
+make nanodbc   # creates shared library
+make tests     # builds the tests
+make test      # runs the tests
+make check     # builds and then runs tests
+make examples  # builds all the example programs
+make install   # installs nanodbc.h and shared library
```
Easier to read.

lexicalunit commented on README.md:
```diff
@@ -278,7 +276,7 @@ Finally, announce the new release to the public.
 - Improve documentation: The main website and API docs should be more responsive.
 - Provide more examples in documentation, more details, and point out any gotchas.
-- Versioned generated source level API documentation for `matesr` and previous releases.
+- Versioned generated source level API documentation for `master` and previous releases.
```
Spelling fix.

lexicalunit commented on README.md:
```diff
@@ -292,9 +290,7 @@ Finally, announce the new release to the public.
 [nanodbc]:              http://nanodbc.io
 [nanodbc-banner]:       https://cloud.githubusercontent.com/assets/1903876/11858632/cc0e21e6-a428-11e5-9a84-39fa27984914.png
 [nanodbc-coverity]:     https://github.com/nanodbc/nanodbc/tree/coverity_scan
-[nanodbc-makefile]:     https://github.com/nanodbc/nanodbc/blob/gh-pages/Makefile
```
Some unused links have been removed from this section.

lexicalunit commented on utility/style.sh:
```diff
@@ -1,5 +1,30 @@
 #!/usr/bin/env bash
-# Source https://github.com/Project-OSRM/osrm-backend
+# Source, modified from: https://github.com/Project-OSRM/osrm-backend
+#
+# LICENSE: BSD-2-Clause
```
Updated from the source, modified, and kept the license.

lexicalunit commented on .github/workflows/ci-linux.yml:
```diff
@@ -174,7 +174,6 @@ jobs:
         cmake --build ${GITHUB_WORKSPACE}/build
 
   test-utility:
-    needs: [build-gcc, build-clang]
```
I've noticed that sometimes jobs in the build matrix fail due to issues not related to nanodbc. For example, sometimes the `actions/checkout@v3` job itself will fail with an error like `libc.so.6: version GLIBC_2.28 not found`. When this happens we skip running all of our tests, because the tests need the _entire_ suite of builds to pass. This gives us an incomplete view of any issues a given PR might introduce. We should run the tests even if some of the build jobs fail.

lexicalunit commented on .github/workflows/ci-windows.yml:
```diff
@@ -106,13 +105,12 @@ jobs:
     - name: Build
       run: |
-        cmake --build ${{ github.workspace }}/build --config Release --target utility_tests 
+        cmake --build ${{ github.workspace }}/build --config Release --target utility_tests
```
Deleting trailing whitespace.

lexicalunit commented on .github/workflows/documentation.yml:
```diff
@@ -18,7 +18,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@master
+    - uses: actions/checkout@v3
```
Using v3 to match other instances of this action used elsewhere.

lexicalunit commented on .github/workflows/ci-linux.yml:
```diff
@@ -60,28 +60,17 @@ jobs:
         - { cxxstd: '14', cxx: g++-9,  cc: gcc-9,  os: ubuntu-20.04 }
         - { cxxstd: '17', cxx: g++-8,  cc: gcc-8,  os: ubuntu-20.04 }
         - { cxxstd: '14', cxx: g++-8,  cc: gcc-8,  os: ubuntu-20.04 }
-        - { cxxstd: '17', cxx: g++-7,  cc: gcc-7,  os: ubuntu-latest, container: "ubuntu:18.04" }
-        - { cxxstd: '14', cxx: g++-7,  cc: gcc-7,  os: ubuntu-latest, container: "ubuntu:18.04" }
-        - { cxxstd: '14', cxx: g++-6,  cc: gcc-6,  os: ubuntu-latest, container: "ubuntu:18.04" }
-        - { cxxstd: '14', cxx: g++-5,  cc: gcc-5,  os: ubuntu-latest, container: "ubuntu:18.04" }
```
`container` is no longer supported: https://github.com/actions/checkout/issues/1474 -- I think these versions are old enough that we can just remove them.

lexicalunit commented on .github/workflows/ci-linux.yml:
```diff
@@ -127,11 +116,6 @@ jobs:
         - { cxxstd: '14', cxx: clang++-8,   cc: clang-8,   os: ubuntu-20.04, pkg: 'libc++-8-dev libc++abi-8-dev' }
         - { cxxstd: '17', cxx: clang++-7,   cc: clang-7,   os: ubuntu-20.04, pkg: 'libc++-7-dev libc++abi-7-dev' }
         - { cxxstd: '14', cxx: clang++-7,   cc: clang-7,   os: ubuntu-20.04, pkg: 'libc++-7-dev libc++abi-7-dev' }
-        - { cxxstd: '17', cxx: clang++-6.0, cc: clang-6.0, os: ubuntu-latest, container: "ubuntu:18.04", pkg: 'libc++-dev libc++abi-dev' }
-        - { cxxstd: '14', cxx: clang++-6.0, cc: clang-6.0, os: ubuntu-latest, container: "ubuntu:18.04", pkg: 'libc++-dev libc++abi-dev' }
-        - { cxxstd: '14', cxx: clang++-5.0, cc: clang-5.0, os: ubuntu-latest, container: "ubuntu:18.04", pkg: 'libc++-dev libc++abi-dev' }
-        - { cxxstd: '14', cxx: clang++-4.0, cc: clang-4.0, os: ubuntu-latest, container: "ubuntu:18.04", pkg: 'libc++-dev libc++abi-dev' }
-        - { cxxstd: '14', cxx: clang++-3.9, cc: clang-3.9, os: ubuntu-latest, container: "ubuntu:18.04", pkg: 'libc++-dev libc++abi-dev' }
```
`container` is no longer supported: https://github.com/actions/checkout/issues/1474 -- I think these versions are old enough that we can just remove them.

lexicalunit commented on .github/workflows/ci-windows.yml:
```diff
@@ -176,20 +173,19 @@ jobs:
       fail-fast: false
       matrix:
-        mysql: [8.0, 5.7]
+        mysql: [8.0]
```
The 5.7 test was failing b/c 5.7 is no longer supported. We can remove it.

lexicalunit commented on .github/workflows/ci-windows.yml:
```diff
@@ -176,20 +173,19 @@ jobs:
         database: nanodbc
         mysql-version: ${{ matrix.mysql }}
-    - name: Install MySQL Connector 8.1
+    - name: Install MySQL Connector
```
This job is actually installing v9.0 here, so the name is misleading.

lexicalunit commented on .github/workflows/ci-windows.yml:
```diff
@@ -204,11 +200,10 @@ jobs:
     - name: Test
       run: |
-        $env:NANODBC_TEST_CONNSTR_MYSQL="Driver={MySQL ODBC 8.3 ANSI Driver};Server=127.0.0.1;Database=nanodbc;User=root;Password=;big_packets=1;NO_SCHEMA=false"
+        $env:NANODBC_TEST_CONNSTR_MYSQL="Driver={MySQL ODBC 8.0 ANSI Driver};Server=127.0.0.1;Database=nanodbc;User=root;Password=;big_packets=1;NO_SCHEMA=false"
```
This test is failing b/c the tests are not finding the driver. I'm attempting to fix this by adjusting the connection string.

lexicalunit commented on .github/workflows/documentation.yml:
```diff
@@ -31,4 +31,4 @@ jobs:
     # TODO: Parse warnings
-    # TODO: Deploy to gh-pages without removing existing folders with docs of historical releases
\ No newline at end of file
+    # TODO: Deploy to gh-pages without removing existing folders with docs of historical releases
```
We should normally always have newlines at the end of files. Just a good practice.

mloskot replied:
I've been using https://pre-commit.com for quite some time, it helps to automate linting and other mundane tasks.

lexicalunit replied:
Yep, I've used that at work for a while now too. Would probably be a good thing to add at some point.

mloskot replied:
https://github.com/nanodbc/nanodbc/issues/417

lexicalunit commented on test/mysql_test.cpp:
```diff
@@ -70,7 +70,7 @@ TEST_CASE_METHOD(mysql_fixture, "test_affected_rows", "[mysql][affected_rows]")
         REQUIRE(result.affected_rows() == 2);
     }
-    // Inseting/retrieving long strings
+    // Inserting/retrieving long strings
```
Spelling fix.

lexicalunit commented on .github/workflows/ci-windows.yml:
```diff
@@ -204,11 +200,10 @@ jobs:
     - name: Test
       run: |
-        $env:NANODBC_TEST_CONNSTR_MYSQL="Driver={MySQL ODBC 8.3 ANSI Driver};Server=127.0.0.1;Database=nanodbc;User=root;Password=;big_packets=1;NO_SCHEMA=false"
+        $env:NANODBC_TEST_CONNSTR_MYSQL="Driver={MySQL ODBC 9.0 ANSI Driver};Server=127.0.0.1;Database=nanodbc;User=root;Password=;big_packets=1;NO_SCHEMA=false"
```
Since `MySQL ODBC 8.3 ANSI Driver` is no longer being installed, this was causing these tests to fail.

lexicalunit commented on .github/workflows/documentation.yml:
```diff
@@ -31,4 +31,4 @@ jobs:
     # TODO: Parse warnings
-    # TODO: Deploy to gh-pages without removing existing folders with docs of historical releases
\ No newline at end of file
+    # TODO: Deploy to gh-pages without removing existing folders with docs of historical releases
```
End of file newline.

lexicalunit commented:
CC @mloskot got all the tests to pass too 🎉

lexicalunit commented:
Bump @mloskot, ok to merge?

mloskot reviewed (approved):
Apologies for delay, LGTM and thanks for this overhaul!
